### DISCUSSION TITLE
Data loader batch multiple graphs

### DISCUSSION THREAD

**Discussion by antoni4040** [pyg-team/pytorch_geometric#3757] [Q&A] [answered]
Hello! I have a basic Dataset class that outputs many small graphs. I'm loading the batches like this:
`graphLoader = DataLoader(graphData, batch_size=BSIZE, sampler=trainSampler)
`
But this seems to flatten the graphs. I saw that there was also the DenseDataLoader, but my graphs have different sizes so it doesn't work. Now, I want to compute a loss based on a specific output from each graph, so I need them to stay separate or else set batch_size to 1. Is there a way to batch them but also keep them separate? Thank you.

**Comment by rusty1s** [comment, marked as answer]
Can you give me an example of what you are trying to achieve? I'm not sure what you mean by "flattening the graph". The graph data shouldn't be flattened at all. The only difference to traditional deep learning mini-batching is that there exists no dedicated batch dimension.

**Reply by antoni4040** [reply]
In my case, I try to predict a value for every edge and I'm interested in the GNN always finding the minimum value, so I try a loss like this:
```
argMin = torch.argmin(a.edge_attr)
loss = ((out.flatten()[argMin]-torch.min(a.edge_attr).item())**2
```

I want to do that for every graph, but if I have a batch size of 12 for example, then torch.min(a.edge_attr) returns the minimum of all those 12 graphs. Unless there's a way to later separate the connected parts of the batch? Sorry if I can't communicate my point, I'm a bit burned out. :sleepy:

**Reply by rusty1s** [reply]
I see. You need to use `torch-scatter` for this:
```python
row, col = edge_index

min_val, min_arg = torch.scatter.scatter_min(edge_attr, batch[row], dim=0)
```

**Reply by antoni4040** [reply]
Interesting library. I write something like this: 
```
for batch in graphLoader:
        optimizer.zero_grad()
        out = model(batch)
        row, col = batch.edge_index
        min_val, min_arg = scatter_min(out, batch[row], dim=0)
```
But I get an index out of bounds error. What exactly is the batch in your code? Again, thank you for your quick reply!

**Reply by antoni4040** [reply]
Oh, sorry, batch.batch in my case I guess, got it. Thanks a lot!
